### PR TITLE
Add decoder hook

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

File: src/avro_binary_decoder.erl [hunk 1]
```diff
@@ -29,79 +29,185 @@
...
+%% A hook fun can be used to fast skipping undesired data fields of records
+%% or undesired data of big maps etc.
+%% For example, to dig out only the field named "myField" in "my.record", the
+%% hook may probably look like:
+%%
+%% fun(Type, SubNameOrIndex, Data, DecodeFun) ->
+%%      case {avro:get_type_fullname(Type), SubNameOrIndex} of
+%%        {"my.record.example.com", "myField"} ->
+%%          DecodeFun(Data);
+%%        {"my.type.example.com", _OtherFields} ->
```
Comment (id) [inline src/avro_binary_decoder.erl]: nitpicking: my.record?

File: src/avro_binary_decoder.erl [hunk 2]
```diff
@@ -29,79 +29,185 @@
...
+    _ = put(avro_binary_decoder_pp_indentation, Indentation + 1),
+    {Result, Tail} = DecodeFun(Data),
+    %% print empty array and empty map
+    case SubInfo =/= [] andalso Result =:= [] of
+      true  -> io:format("~s  []\n", [IndentationStr]);
+      false -> ok
+    end,
+    %% print the value if it's a leaf in the type tree
+    _ = SubInfo =:= [] andalso io:format("~1000000p\n", [Result]),
+    _ = put(avro_binary_decoder_pp_indentation, Indentation),
```
Comment (id) [inline src/avro_binary_decoder.erl]: Why reset avro_binary_decoder_pp_indentation back to Indentation? On line 136 you set it to Indentation + 1.

Comment (zmstone) [inline src/avro_binary_decoder.erl]: +1 before going deep into the stack, and set back after popped back up to the current depth in the stack.

Comment (id) [inline src/avro_binary_decoder.erl]: Right.

**Comment by id** [comment]
Very nice.

**Pull request merged**